### PR TITLE
Introduce INC3.0 auto-tune

### PR DESCRIPTION
**Pull request by yiliu30**
## Type of Change

Feature
API changed: Introduce the auto-tune API

## Description

As mentioned in [The design for autotune in INC 3.0](https://github.com/intel/neural-compressor/discussions/1445), this is the 1st PR to introduce `autotune` in INC 3.0, which includes an overview of the auto-tune pipeline and main interface.

Please refer to [UTs](https://github.com/intel/neural-compressor/blob/8f397b706b2223fdb56dea47685f34a8c225eda4/test/3x/torch/test_autotune.py#L43-L114) for the usage.

## How has this PR been tested?

Pre-CI

## Dependency Change?

None

### CODE REVIEW

File: neural_compressor/common/base_tune.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    def __init__(self, model) -> None:
+        self.model = model
+
+    @abstractmethod
+    def apply(self):
+        """The entry to apply algorithms on a given model."""
+        raise NotImplementedError
+
+
+class TuningObjective:
```
Comment (ftian1) [inline neural_compressor/common/base_tune.py:41]: shall we explicitly distinguish single objective and multiple objectives? if not, I would suggest to rename as TuningObjectives

File: neural_compressor/common/base_tune.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            The overall result of all registered evaluation functions.
+        """
+        result = 0
+        for eval_pair in self.eval_fn_registry:
+            eval_fn = eval_pair[self.EVAL_FN]
+            eval_result = eval_fn(model)
+            result = self._update_the_objective_score(eval_pair, eval_result, result)
+        return result
+
+    def _update_the_objective_score(self, eval_pair, eval_result, overall_result):
```
Comment (ftian1) [inline neural_compressor/common/base_tune.py:66]: coding style consistent, some of funcs like line 50 has return value declaration, but here it hasn't.

File: neural_compressor/common/base_tune.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        self.eval_fn_registry = [
+            {
+                self.EVAL_FN: user_eval_fn_pair[self.EVAL_FN],
+                self.WEIGHT: user_eval_fn_pair.get(self.WEIGHT, 1.0),
+                self.FN_NAME: user_eval_fn_pair.get(self.FN_NAME, user_eval_fn_pair[self.EVAL_FN].__name__),
+            }
+            for user_eval_fn_pair in user_eval_fns
+        ]
+
+    def update_eval_fn_registry(self, eval_fns: Optional[Union[Dict, List[Dict]]] = None) -> None:
```
Comment (ftian1) [inline neural_compressor/common/base_tune.py:83]: shall rename this func to a `set` method as it's used to register objectives, right?

File: neural_compressor/common/base_tune.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        elif isinstance(eval_fns, Dict):
+            eval_fns = [eval_fns]
+        elif isinstance(eval_fns, List):
+            assert all([isinstance(eval_fn_pair, Dict) for eval_fn_pair in eval_fns])
+        else:
+            raise NotImplementedError(f"The eval_fns should be a dict or a list of dict, but got {type(eval_fns)}.")
+        self._update_eval_fn_registry(eval_fns)
+
+
+tuning_objective = TuningObjective()
```
Comment (ftian1) [inline neural_compressor/common/base_tune.py:95]: define a default objective for use?

File: neural_compressor/common/base_tune.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    """
+
+    def __init__(self, quant_configs=None, timeout=0, max_trials=100):
+        """Init a TuningCriterion object."""
+        self.quant_configs = quant_configs
+        self.timeout = timeout
+        self.max_trials = max_trials
+
+
+class Tuner:
```
Comment (ftian1) [inline neural_compressor/common/base_tune.py:114]: Better to use `Tuning`?

File: neural_compressor/common/base_tune.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    def generate_quant_config(quant_config: BaseConfig) -> List[BaseConfig]:
+        if isinstance(quant_config, ComposableConfig):
+            result = []
+            for q_config in quant_config.config_list:
+                result += q_config.expand()
+            return result
+        else:
+            return quant_config.expand()
+
+    def generate_quant_config_from_quant_configs(self):
```
Comment (ftian1) [inline neural_compressor/common/base_tune.py:138]: same concern

File: neural_compressor/common/utility.py [hunk 7]
```diff
@@ -36,3 +36,10 @@
 from typing import Callable, Union
 
 OP_NAME_OR_MODULE_TYPE = Union[str, Callable]
+
+
+def convert_to_list(value):
```
Comment (ftian1) [inline neural_compressor/common/utility.py:41]: such util looks too naive...

File: neural_compressor/torch/tune.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        super().__init__(model)
+        self.run_fn = run_fn
+        self.run_args = run_args
+
+    def apply(self, quant_config):
+        """The entry to apply quantization algorithms on a given a model."""
+        logger.info(f"apply quant_config: {quant_config}.")
+        from neural_compressor.torch import quantize
+
+        q_model = quantize(model=self.model, quant_config=quant_config, run_fn=self.run_fn, run_args=self.run_args)
```
Comment (ftian1) [inline neural_compressor/torch/tune.py:46]: is this just an example and will be changed later? why torch algo must invoke quantize() to apply?

File: neural_compressor/torch/tune.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+def autotune(
+    model: torch.nn.Module,
+    tune_config: TuningConfig,
+    eval_fns: Optional[Union[Dict, List[Dict]]] = None,
+    run_fn=None,
+    run_args=None,
+):
+    tuning_objective.update_eval_fn_registry(eval_fns)
+    tuner = Tuner(tune_config=tune_config)
```
Comment (ftian1) [inline neural_compressor/torch/tune.py:63]: why `objective` is not one of input parameters of Tuner class?

File: neural_compressor/torch/tune.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    algo_manager = TorchAlgorithmWrapper(model, run_fn, run_args)
```
Comment (ftian1) [inline neural_compressor/torch/tune.py:64]: I don't understand these two lines usage quite well.

File: neural_compressor/common/base_tune.py (hunk 1 above)
Comment (yiliu30) [reply neural_compressor/common/base_tune.py:41]: We don't explicitly distinguish between single and multiple objectives. Renamed it to `TuningObjectives`.

File: neural_compressor/common/base_tune.py (hunk 3 above)
Comment (yiliu30) [reply neural_compressor/common/base_tune.py:83]: Yes, renamed it to `set_xxx`.

File: neural_compressor/common/base_tune.py (hunk 2 above)
Comment (yiliu30) [reply neural_compressor/common/base_tune.py:66]: Thanks, added.

File: neural_compressor/common/base_tune.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+    def _post_init(self):
+        # check the number of evaluation functions
+        num_tuning_objectives = self.tuner_tuning_objective.get_number_of_tuning_objectives()
+        assert (
+            num_tuning_objectives > 0
+        ), "Please ensure that you register at least one evaluation metric for auto-tune."
+        logger.info(f"There are {num_tuning_objectives} tuning objectives.")
+
+    @staticmethod
+    def generate_quant_config(quant_config: BaseConfig) -> List[BaseConfig]:
```
[the remark this answers is not in the document]
Comment (yiliu30) [reply neural_compressor/common/base_tune.py:129]: Is okay to rename it to `parse_quant_config`?

File: neural_compressor/common/utility.py (hunk 7 above)
Comment (yiliu30) [reply neural_compressor/common/utility.py:41]: Updated.

File: neural_compressor/torch/tune.py (hunk 8 above)
Comment (yiliu30) [reply neural_compressor/torch/tune.py:46]: Yes, I think we can refine this interface after we unify the algorithms' interfaces.

File: neural_compressor/torch/tune.py (hunk 9 above)
Comment (yiliu30) [reply neural_compressor/torch/tune.py]: Refactor the code to use `objective` and `xx_wrapper` as the input args of `Tuner`.

File: neural_compressor/torch/tune.py (hunk 10 above)
Comment (yiliu30) [reply neural_compressor/torch/tune.py]: Refactor the code to utilize `objective` and `xx_wrapper` as the input args for Tuner.

File: neural_compressor/common/base_tune.py (hunk 4 above)
Comment (yiliu30) [reply neural_compressor/common/base_tune.py]: The `TuningObjectives` comprises user-provided eval_fn, and there isn't a default value for it. The `TuningConfig` comes with a default value and will be added in the next PR.

**Review by zehao-intel — APPROVED** [review]

**Review by xin3he — APPROVED** [review]

**Review by mengniwang95 — APPROVED** [review]

File: neural_compressor/common/base_tune.py (hunk 5 above)
Comment (yiliu30) [reply neural_compressor/common/base_tune.py:113]: On GitHub, the `class Tuner` is more frequently used than the `class Tuning`.

**Pull request merged**